synapse/storage/schema/main/delta/73/06thread_notifications_thread_id_idx.sql

```diff
@@ -13,7 +13,11 @@
  * limitations under the License.
  */
 
--- The columns can now be made non-nullable.
-ALTER TABLE event_push_actions_staging ALTER COLUMN thread_id SET NOT NULL;
-ALTER TABLE event_push_actions ALTER COLUMN thread_id SET NOT NULL;
-ALTER TABLE event_push_summary ALTER COLUMN thread_id SET NOT NULL;
+-- Allow there to be multiple summaries per user/room.
+DROP INDEX IF EXISTS event_push_summary_unique_index;
+
+INSERT INTO background_updates (ordering, update_name, progress_json, depends_on) VALUES
+  (7306, 'event_push_actions_thread_id_null', '{}', 'event_push_backfill_thread_id');
+
+INSERT INTO background_updates (ordering, update_name, progress_json, depends_on) VALUES
+  (7306, 'event_push_summary_thread_id_null', '{}', 'event_push_backfill_thread_id');
```

Comment:
I chose not to care about `event_push_actions_staging` here since it is transient and should be handled OK once the rows get into `event_push_actions`.

Reply:
> Though at this point it might be worth just doing the UPDATE when we do the others too

But we don't need to since we don't join to the table; we could maybe do something when we pull rows from `event_push_actions_staging` and insert into `event_push_actions`? But the data can only be "incorrect" across a restart of Synapse, I believe, which ends up orphaning the data anyway (since the event wouldn't be persisted).

Reply:
> But the data can only be "incorrect" across a restart of Synapse, I believe, which ends up orphaning the data anyway (since the event wouldn't be persisted).

My only mild concern is what happens when we do a rolling restart or something, but I think that's fine?

Reply:
> > But the data can only be "incorrect" across a restart of Synapse, I believe, which ends up orphaning the data anyway (since the event wouldn't be persisted).
> 
> My only mild concern is what happens when we do a rolling restart or something, but I think that's fine?

(Note that Synapse 1.69 fills in that column in the staging table so shouldn't be an issue)

I don't think it will be a problem regardless:

1. `event_push_actions_staging` has a null `thread_id` value.
2. It gets moved over to `event_push_actions` (there's now a null in `event_push_actions`).
3. Before we use `event_push_actions` the column gets updated to be `"main"` and things work.

There might be a slight race in here with whether bg update has ran though. 😟